File: services/src/main/java/com/senacor/code/fullstack/chat/rest/ChatMessageController.java
package com.senacor.code.fullstack.chat.rest;

import com.senacor.code.fullstack.chat.domain.ChatMessage;
import com.senacor.code.fullstack.chat.service.ChatMessageService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * The REST controller for managing chat messages.
 */
@RestController
@RequestMapping("/api/v1/channels/{channel}/messages")
public class ChatMessageController {

    private ChatMessageService messageService;

    @Autowired
    public ChatMessageController(ChatMessageService messageService) {
        this.messageService = messageService;
    }

    @GetMapping
    public List<ChatMessage> loadChatMessages(@PathVariable("channel") String channel) {
        return messageService.loadChatMessages(channel);
    }
}

File: services/src/main/kotlin/com/senacor/code/fullstack/chat/message/ChatMessageController.kt
package com.senacor.code.fullstack.chat.message

import org.springframework.web.bind.annotation.*
import org.springframework.web.util.UriComponentsBuilder
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.http.HttpHeaders



@RestController
@RequestMapping("/api/channels/{channelId}/messages")
class ChatMessageController(private val messageService: ChatMessageService) {

    @GetMapping
    fun loadMessages(@PathVariable("channelId") channelId: String) = messageService.loadMessages(channelId)

    @PostMapping
    fun saveMessage(@PathVariable("channelId") channelId: String, sender: String, message: String): ResponseEntity<Void> {
        val msg = messageService.createMessage(channelId, sender, message)
        val headers = HttpHeaders()
        headers.location = UriComponentsBuilder.newInstance()
                .path("/api/channels/{channelId}/messages/{messageId}")
                .buildAndExpand(msg.channelId, msg.messageId).toUri()
        return ResponseEntity<Void>(headers, HttpStatus.CREATED)
    }

}